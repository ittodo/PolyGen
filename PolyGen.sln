Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio Version 17
VisualStudioVersion = 17.5.2.0
MinimumVisualStudioVersion = 10.0.40219.1
Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "dist", "dist", "{AC15D111-0E66-69FF-AB22-C1C2B24C9B2F}"
EndProject
Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "csharp-gen-demo", "csharp-gen-demo", "{50ED58CA-0D75-0E21-8EE7-A78D748B2F09}"
EndProject
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "GenDemo", "dist\csharp-gen-demo\GenDemo\GenDemo.csproj", "{E226D435-B4C7-34D6-09C0-7130806E34EB}"
EndProject
Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "jsoncsv-demo", "jsoncsv-demo", "{36589BE5-A03F-449E-102D-01BABBA6554E}"
EndProject
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "JsonCsvDemo", "dist\jsoncsv-demo\JsonCsvDemo\JsonCsvDemo.csproj", "{A62150AD-B712-B2A5-3016-F4DBAE33D403}"
EndProject
Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "run-csharp", "run-csharp", "{094DD1BF-3AE8-C1B3-4526-AA95770202DE}"
EndProject
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "RunDemo", "dist\run-csharp\RunDemo\RunDemo.csproj", "{F8FB6063-A8F5-6964-28EE-22701C1C3D29}"
EndProject
Global
	GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Debug|Any CPU = Debug|Any CPU
		Release|Any CPU = Release|Any CPU
	EndGlobalSection
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
		{E226D435-B4C7-34D6-09C0-7130806E34EB}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{E226D435-B4C7-34D6-09C0-7130806E34EB}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{E226D435-B4C7-34D6-09C0-7130806E34EB}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{E226D435-B4C7-34D6-09C0-7130806E34EB}.Release|Any CPU.Build.0 = Release|Any CPU
		{A62150AD-B712-B2A5-3016-F4DBAE33D403}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{A62150AD-B712-B2A5-3016-F4DBAE33D403}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{A62150AD-B712-B2A5-3016-F4DBAE33D403}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{A62150AD-B712-B2A5-3016-F4DBAE33D403}.Release|Any CPU.Build.0 = Release|Any CPU
		{F8FB6063-A8F5-6964-28EE-22701C1C3D29}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{F8FB6063-A8F5-6964-28EE-22701C1C3D29}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{F8FB6063-A8F5-6964-28EE-22701C1C3D29}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{F8FB6063-A8F5-6964-28EE-22701C1C3D29}.Release|Any CPU.Build.0 = Release|Any CPU
	EndGlobalSection
	GlobalSection(SolutionProperties) = preSolution
		HideSolutionNode = FALSE
	EndGlobalSection
	GlobalSection(NestedProjects) = preSolution
		{50ED58CA-0D75-0E21-8EE7-A78D748B2F09} = {AC15D111-0E66-69FF-AB22-C1C2B24C9B2F}
		{E226D435-B4C7-34D6-09C0-7130806E34EB} = {50ED58CA-0D75-0E21-8EE7-A78D748B2F09}
		{36589BE5-A03F-449E-102D-01BABBA6554E} = {AC15D111-0E66-69FF-AB22-C1C2B24C9B2F}
		{A62150AD-B712-B2A5-3016-F4DBAE33D403} = {36589BE5-A03F-449E-102D-01BABBA6554E}
		{094DD1BF-3AE8-C1B3-4526-AA95770202DE} = {AC15D111-0E66-69FF-AB22-C1C2B24C9B2F}
		{F8FB6063-A8F5-6964-28EE-22701C1C3D29} = {094DD1BF-3AE8-C1B3-4526-AA95770202DE}
	EndGlobalSection
	GlobalSection(ExtensibilityGlobals) = postSolution
		SolutionGuid = {5A8F19CD-26E9-43DF-B64B-C40265ED6023}
	EndGlobalSection
EndGlobal
