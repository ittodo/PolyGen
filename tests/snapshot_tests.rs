use anyhow::Result;
use insta::assert_debug_snapshot;
use PolyGen::{parse_and_merge_schemas, build_ir_from_asts};
use walkdir::WalkDir;

#[test]
fn test_ast_snapshots() -> Result<()> {
    for entry in WalkDir::new("tests/schemas")
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| e.path().extension().map_or(false, |ext| ext == "poly"))
    {
        let schema_path = entry.path();
        let schema_name = schema_path.file_stem().unwrap().to_str().unwrap();

        // Test AST generation
        let all_asts = parse_and_merge_schemas(schema_path, None)?; // Pass None for output_dir in tests
        assert_debug_snapshot!(format!("{}_ast", schema_name), all_asts);

        // Test IR generation
        let ir_context = build_ir_from_asts(&all_asts);
        assert_debug_snapshot!(format!("{}_ir", schema_name), ir_context);
    }
    Ok(())
}