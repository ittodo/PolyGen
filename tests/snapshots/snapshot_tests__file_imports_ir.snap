---
source: tests/snapshot_tests.rs
expression: ir_context
---
SchemaContext {
    files: [
        FileDef {
            path: "file_imports.poly",
            namespaces: [
                NamespaceDef {
                    name: "",
                    items: [
                        Struct(
                            StructDef {
                                name: "MainTable",
                                fqn: "MainTable",
                                header: [],
                                items: [
                                    Field(
                                        FieldDef {
                                            name: "id",
                                            field_type: TypeRef {
                                                original: "u32",
                                                fqn: "u32",
                                                namespace_fqn: "",
                                                type_name: "u32",
                                                lang_type: "u32",
                                                is_primitive: true,
                                                is_struct: false,
                                                is_enum: false,
                                                is_option: false,
                                                is_list: false,
                                                inner_type: None,
                                            },
                                            attributes: [],
                                        },
                                    ),
                                    Field(
                                        FieldDef {
                                            name: "ref_a",
                                            field_type: TypeRef {
                                                original: "ImportedA",
                                                fqn: "ImportedA",
                                                namespace_fqn: "",
                                                type_name: "ImportedA",
                                                lang_type: "ImportedA",
                                                is_primitive: false,
                                                is_struct: true,
                                                is_enum: false,
                                                is_option: false,
                                                is_list: false,
                                                inner_type: None,
                                            },
                                            attributes: [],
                                        },
                                    ),
                                    Field(
                                        FieldDef {
                                            name: "status_b",
                                            field_type: TypeRef {
                                                original: "StatusB",
                                                fqn: "StatusB",
                                                namespace_fqn: "",
                                                type_name: "StatusB",
                                                lang_type: "StatusB",
                                                is_primitive: false,
                                                is_struct: false,
                                                is_enum: true,
                                                is_option: false,
                                                is_list: false,
                                                inner_type: None,
                                            },
                                            attributes: [],
                                        },
                                    ),
                                ],
                            },
                        ),
                    ],
                },
            ],
        },
        FileDef {
            path: "imported_schema_a.poly",
            namespaces: [
                NamespaceDef {
                    name: "",
                    items: [
                        Struct(
                            StructDef {
                                name: "ImportedA",
                                fqn: "ImportedA",
                                header: [],
                                items: [
                                    Field(
                                        FieldDef {
                                            name: "id",
                                            field_type: TypeRef {
                                                original: "u32",
                                                fqn: "u32",
                                                namespace_fqn: "",
                                                type_name: "u32",
                                                lang_type: "u32",
                                                is_primitive: true,
                                                is_struct: false,
                                                is_enum: false,
                                                is_option: false,
                                                is_list: false,
                                                inner_type: None,
                                            },
                                            attributes: [],
                                        },
                                    ),
                                    Field(
                                        FieldDef {
                                            name: "value",
                                            field_type: TypeRef {
                                                original: "string",
                                                fqn: "string",
                                                namespace_fqn: "",
                                                type_name: "string",
                                                lang_type: "string",
                                                is_primitive: true,
                                                is_struct: false,
                                                is_enum: false,
                                                is_option: false,
                                                is_list: false,
                                                inner_type: None,
                                            },
                                            attributes: [],
                                        },
                                    ),
                                ],
                            },
                        ),
                    ],
                },
            ],
        },
        FileDef {
            path: "imported_schema_b.poly",
            namespaces: [
                NamespaceDef {
                    name: "",
                    items: [
                        Enum(
                            EnumDef {
                                name: "StatusB",
                                fqn: "StatusB",
                                items: [
                                    Member(
                                        EnumMember {
                                            name: "Enabled",
                                            value: Some(
                                                0,
                                            ),
                                        },
                                    ),
                                    Member(
                                        EnumMember {
                                            name: "Disabled",
                                            value: Some(
                                                1,
                                            ),
                                        },
                                    ),
                                ],
                            },
                        ),
                    ],
                },
            ],
        },
    ],
}
